### PR TITLE
Support Symlinks for cached download archives

### PR DESCRIPTION
**Pull request by andrew-edgar**
Added support for Symlinks in buildpacks and any other cached downloaded artifacts

Signed-off-by: Roberto Jimenez <roberto@de.ibm.com>

### CODE REVIEW

**Comment by ematpl** [comment]
Thanks, @andrew-edgar! Prioritizing for the Diego team to pull in. With regard to your question on https://github.com/cloudfoundry/cacheddownloader/issues/12, the Diego team did do https://www.pivotaltracker.com/story/show/88934110 to address processing of symlinks elsewhere in the cacheddownloader. I think it was to support the PHP buildpack correctly, but that may have been a much older version of that buildpack.

**Comment by andrew-edgar** [comment]
I just found that my change is not quite right in that it doesn't fix the problems we saw with our buildpack.  I will be submitting a small update to this PR that does fix the problem shortly

File: file_cache_test.go [hunk 1]
```diff
@@ -752,11 +762,88 @@ var _ = Describe("FileCache", func() {
...
+
+				Context("and the directory is later fetched", func() {
+					JustBeforeEach(func() {
+						dir, cacheInfoType, getErr = cache.GetDirectory(cacheKey)
+						Expect(getErr).NotTo(HaveOccurred())
+						Expect(dir).NotTo(BeEmpty())
+						Expect(cacheInfoType).To(Equal(cacheInfo))
+					})
+
+					It("returns a valid path", func() {
```
Comment (andrew-edgar) [inline file_cache_test.go]: Sorry yeah I merged some code in and there are a few tests that got added again accidentily.  Did you want to just fix them or want me to update and resubmit?

**Comment by andrew-edgar** [comment]
I updated the tests to fix all that stuff that was added incorrectly.   Sorry about that :)

I like the idea of using the extractor package instead of us doing it.

**Pull request closed without merging**